Add participation plugin

alexsporn opened the pull request:
This plugin enables voting on the tangle.

muXxer committed:
Initial draft for referendum plugin

alexsporn committed:
- Remove the Question index field

alexsporn committed:
- Added Votes type that can hold the votes for multiple referenda

alexsporn committed:
- Adapted ReferendumManager and plugin to handle votes

alexsporn committed:
- Store the messages for each vote and track the current vote for each referendum

alexsporn committed:
- Add Referendum plugin to list of plugins

alexsporn committed:
- Add database initialization and closing to the Referendum plugin

alexsporn committed:
- When adding a new referendum, verify that it hasn't already started or ended

alexsporn committed:
- Store for each output the start and end milestone when it was counted and the message that generated it

alexsporn committed:
- Added endpoint to query the vote status of a given outputID

alexsporn committed:
- Add optional indexation data to the testsuite MessageBuilder

- Adapted TestWhiteFlagSendAllCoins to issue and assert a milestone in between

alexsporn committed:
- Add optional callbacks to the testsuite that emulate the Tangle pkg events triggered on milestone confirmation

alexsporn committed:
- Add a QuestionBuilder to the referendum package

alexsporn committed:
- Added save, load and delete  referendums to the store

alexsporn committed:
- Fixed Answer json marshalling

alexsporn committed:
- Add SyncManager dependency to the ReferendumManager to get the current confirmed milestone index and do the proper checks

alexsporn committed:
- Made the last milestone messageId public in the testsuite

alexsporn committed:
- Added a Votes builder

alexsporn committed:
- Added a test setup for the ReferendumManager and an initial test verifying the Referendum states according to the confirmedMilestoneIndex

alexsporn committed:
- Added serialization validation for Question and Referendum

alexsporn committed:
- Move the ReferendumStatus type and logic from the plugin to the pkg

alexsporn committed:
- Fixed state calculation and added tests checking the filter functions

alexsporn committed:
- Add first test voting and counting votes

alexsporn committed:
- Assert, that the ReferendumManager holds the message of a vote in its storage

alexsporn committed:
- Use correct prefix to store referendums

alexsporn committed:
- Refactored referendum test utilities into a separate package

alexsporn committed:
- Added tests cancelling referendum votes

- Fixed a bug found with the new tests

alexsporn committed:
- Separated the active and past votes into separate prefixes so we can query them

alexsporn committed:
- Track each vote individually for each referendum

- Added the amount used to the tracked vote

alexsporn committed:
- End all tracked votes when the referendum ends

alexsporn committed:
- Added new status to the Referendum for the API with `upcoming`, `commencing`, `holding` and `ended`

alexsporn committed:
- Some renamed in both structs and JSON keys

alexsporn committed:
- Migrated codebase to new iota.go and hive.go and now using serializer pkg from hive.go

alexsporn committed:
- Migrated referendum pkg to use hive.go serializable

alexsporn committed:
- Add variable payloads to referendum

alexsporn committed:
- Renaming and refactoring stuff to make it more consistent

alexsporn committed:
- Renamed Questions to Ballot

alexsporn committed:
- Added a new VoteBuilder to the test environment

alexsporn committed:
- Refactored test code to avoid so much duplication

alexsporn committed:
- Add test with multiple vote outputs

alexsporn committed:
- Add test with multiple votes to with different answers

alexsporn committed:
- Add test where an opinion changed

alexsporn committed:
- Added test with concurrent referendums and catched 2 bugs that were caused by early voting

alexsporn committed:
- Added type to ballot json representation

alexsporn committed:
- Remove referendum db.path config in favour of creating a subdir in the main db folder

alexsporn committed:
- Added health tracker for the referendum database

alexsporn committed:
- Added empty test files for all the tests that still need to be added

alexsporn committed:
- Fixed tests still using iota.go instead of hive.go for serialization

alexsporn committed:
- Rename milestoneIndex fields of referendum.

- Removed some TODOs and fixed some comments

alexsporn committed:
- Rename referendum pkg to participation

alexsporn committed:
- Renamed ReferendumTestEnv to ParticipationTestEnv

alexsporn committed:
- Rename ReferendumManager to ParticipationManager

alexsporn committed:
- Rename milestoneIndexBeginHolding to milestoneIndexStart

alexsporn committed:
- Remove params from plugin

alexsporn committed:
- Renamed Referedum plugin to Participation

alexsporn committed:
- Renaming Referendum to ParticipationEvent

alexsporn committed:
- Fixed typo in pkg name

alexsporn committed:
- Rename ReferendumBuilder to ParticipationEventBuilder

alexsporn committed:
- Renaming files

alexsporn committed:
- More renaming and adapting the rest API

alexsporn committed:
- More renaming

alexsporn committed:
- Further renames

alexsporn committed:
- Add optional type query parameter to the events endpoint

alexsporn committed:
- fixed review dog bug

alexsporn committed:
- Review doggo fixes

alexsporn committed:
- Added some documentation and updated TODOs

muXxer commented on pkg/model/storage/health.go:
```diff
@@ -74,22 +74,22 @@ func (s *storeHealthTracker) setDatabaseVersion(version byte) error {
 
 	if len(value) > 0 {
-		return value[0] == expectedVersion, nil
+		return value[0] == DBVersion, nil
```
So we have one global version for all different databases? Does that make sense?

alexsporn replied:
Thing is we currently do not have a need for versioning and migrations. I'd keep it like that until we really need to migrate something

muXxer replied:
alright, was just wondering because why should the main database version be increased, if we change something for example in the participation database layout :)

muXxer commented on pkg/shutdown/shutdown.go:
```diff
@@ -28,8 +28,9 @@ const (
-	PriorityFaucet  // depends on PriorityPoWHandler
+	PrioritySpammer    // depends on PriorityPoWHandler
+	PriorityFaucet     // depends on PriorityPoWHandler
+	PriorityReferendum // depends on PriorityPoWHandler
```
`PriorityParticipation`

muXxer commented on pkg/testsuite/test_environment.go:
```diff
@@ -79,8 +79,21 @@ type TestEnvironment struct {
+	// OnNewSpent callback that will be called for each spent UTXO. This is equivalent to the tangle.NewUTXOSpent event.
+	OnNewSpent OnNewSpentFunc
+
+	// OnNewSpent callback that will be called after confirming a milestone. This is equivalent to the tangle.ConfirmedMilestoneIndexChanged event.
```
Wrong comment `OnNewSpent`

muXxer commented on pkg/whiteflag/test/white_flag_test.go:
```diff
@@ -66,9 +77,9 @@ func TestWhiteFlagSendAllCoins(t *testing.T) {
-	require.Equal(t, 2+1, confStats.MessagesReferenced) // 2 + milestone itself
-	require.Equal(t, 2, confStats.MessagesIncludedWithTransactions)
+	_, confStats = te.IssueAndConfirmMilestoneOnTips(hornet.MessageIDs{messageB.StoredMessageID()}, true)
+	require.Equal(t, 1+1, confStats.MessagesReferenced) // 2 + milestone itself
```
wrong comment

muXxer commented on pkg/testsuite/utils.go:
```diff
@@ -42,13 +44,21 @@ type Message struct {
 }
 
+func (b *MessageBuilder) LatestMilestonesAsParents() *MessageBuilder {
+	return b.Parents(hornet.MessageIDs{b.te.Milestones[len(b.te.Milestones)-1].Milestone().MessageID, b.te.Milestones[len(b.te.Milestones)-2].Milestone().MessageID})
```
Should we check if the `b.te.Milestones` has elements? Or do we ignore that because testsuite anyway?

alexsporn replied:
We normally initialize it with at least 2-3 milestones. Since this is for tests I would skip the checks.

muXxer replied:
yeah right, I also saw that later after adding the comment :)

muXxer commented on pkg/model/participation/database_prefixes.go:
```diff
@@ -0,0 +1,17 @@
+	ParticipationStoreKeyPrefixMessages byte = 1
+
+	// Tracks all active and past participations
+	ParticipationStoreKeyPrefixReferendumOutputs      byte = 2
```
Referendum?

muXxer commented on pkg/model/participation/ballot_builder.go:
```diff
@@ -0,0 +1,33 @@
+	return qb
+}
+
+// Build builds the Question.
```
the ballot

muXxer commented on pkg/model/participation/ballot_builder.go:
```diff
@@ -0,0 +1,33 @@
+	ballot *Ballot
+}
+
+// AddQuestion adds the given question to the questions.
```
to the ballot questions

muXxer commented on pkg/model/participation/tracked_participation.go:
```diff
@@ -0,0 +1,100 @@
+func (t *TrackedParticipation) valueBytes() []byte {
+	m := marshalutil.New(48)
+	m.WriteBytes(t.MessageID) // 32 bytes
+	m.WriteUint64(t.Amount)
```
comment // 8 bytes

muXxer commented on plugins/participation/participation.go:
```diff
@@ -0,0 +1,193 @@
+	}
+
+	if len(b) != participation.EventIDLength {
+		return participation.EventID{}, fmt.Errorf("unknown eventID length (%d)", len(b))
```
Why not `participation.NullEventID` here as well?

muXxer commented on pkg/model/participation/test/testenv.go:
```diff
@@ -0,0 +1,328 @@
+
+	store := mapdb.NewMapDB()
+
+	rm, err := participation.NewManager(
```
maybe rename to `pm`since it is a participation manager now

muXxer commented on pkg/model/participation/test/testenv.go:
```diff
@@ -0,0 +1,328 @@
+	question, err := questionBuilder.Build()
+	require.NoError(env.t, err)
+
+	questionsBuilder := participation.NewBallotBuilder()
```
ballotBuilder

muXxer commented on pkg/model/participation/event.go:
```diff
@@ -0,0 +1,272 @@
+}
+
+func (e *Event) UnmarshalJSON(bytes []byte) error {
+	jReferendum := &jsonEvent{}
```
jEvent

muXxer commented on pkg/model/participation/participation_manager.go:
```diff
@@ -0,0 +1,597 @@
+	}
+}
+
+// WithLogger enables logging within the faucet.
```
hups, that was me :D
Its not the faucet

muXxer commented on pkg/model/participation/participation_manager.go:
```diff
@@ -0,0 +1,597 @@
+	}
+}
+
+// Option is a function setting a faucet option.
```
same here :weirdsmileyguy:

muXxer commented on pkg/model/participation/participation_manager.go:
```diff
@@ -0,0 +1,597 @@
+	return filtered
+}
+
+// EventsAcceptingParticipation returns the events that are currently accepting participation, i.event. commencing or in the holding period.
```
What does `i.event.` mean?

muXxer commented on pkg/model/participation/participation_manager.go:
```diff
@@ -0,0 +1,597 @@
+	})
+}
+
+// EventsCountingParticipation returns the events that are currently actively counting participation, i.event. in the holding period
```
What does `i.event.` mean?

muXxer commented on pkg/model/participation/participation_manager.go:
```diff
@@ -0,0 +1,597 @@
+// 	- Output Type 0 (SigLockedSingleOutput) and Type 1 (SigLockedDustAllowanceOutput) are both valid for this.
+// 	- The Indexation must match the configured Indexation.
+//  - The participation data must be parseable.
+func (rm *ParticipationManager) ApplyNewUTXO(index milestone.Index, newOutput *utxo.Output) error {
```
Rename all occurrences of ParticipationManager to `pm`

muXxer commented on pkg/model/participation/participation_manager.go:
```diff
@@ -0,0 +1,597 @@
+		for idx, question := range event.BallotQuestions() {
+			questionIndex := uint8(idx)
+
+			// For each question, iterate over all answers. Include 0 here, since that is valid, i.event. answer skipped by voter
```
`i.event.`

muXxer commented on pkg/model/participation/participation_manager_test.go:
```diff
@@ -0,0 +1,758 @@
+
+	env.IssueMilestone() // 7
+
+	// Event should be ended
```
wrong comment

muXxer commented on pkg/model/participation/participations_builder.go:
```diff
@@ -0,0 +1,35 @@
+// Build builds the Participations.
+func (b *ParticipationsBuilder) Build() (*Participations, error) {
+	if _, err := b.p.Serialize(serializer.DeSeriModePerformValidation); err != nil {
+		return nil, fmt.Errorf("unable to build question: %w", err)
```
participation

muXxer commented on pkg/model/participation/test/builder.go:
```diff
@@ -0,0 +1,106 @@
+	return b
+}
+
+func (b *ParticipationHelper) AddDefaultBallotVote(eventID participation.EventID) *ParticipationHelper {
```
Is this function name still correct, or is this a `AddDefaultParticipation` ?

alexsporn replied:
Yes, it explicitely adds a vote for a ballot

alexsporn committed:
- Fixed review comments

muXxer reviewed (commented):
LGTM for now :+1:

alexsporn committed:
- Added some TODOs

alexsporn committed:
- Separate the add/remove event API operations to an admin endpoint that can be protected by JWT by the node operator

alexsporn committed:
- Added more validation for ballot and event

alexsporn committed:
- Renamed Answer.Index to Answer.Value

- Map the given answer to either a valid one or to invalid (255)
- Reworked the vote counting to not work with answer indexes but use the values
- Added validation logic to the Question to avoid having duplicate Answer values
- Divided the voting amount by 1000 to calculate the amount of votes

muXxer reviewed (commented):
Changes look good.

muXxer commented on pkg/model/participation/question.go:
```diff
@@ -134,3 +172,26 @@ func (j *jsonQuestion) ToSerializable() (serializer.Serializable, error) {
 	return payload, nil
 }
+
+// QuestionAnswers returns the possible answers for a Que
```
There is some text missing at the end.

muXxer commented on pkg/model/participation/question.go:
```diff
@@ -134,3 +172,26 @@ func (j *jsonQuestion) ToSerializable() (serializer.Serializable, error) {
+	return answers
+}
+
+// answerValueForByte checks if the given value is a valid and maps any other values to AnswerValueInvalid
```
is a valid what?

muXxer reviewed (commented):
Index to Value changes LGTM

alexsporn committed:
- Fixed 2 comments

alexsporn committed:
- Add a checksum to the event status to easily compare if the results match

alexsporn committed:
- Added Serialization/Deserialization tests for Question and Answer

alexsporn committed:
- Added TrackedParticipation serialization tests

alexsporn committed:
- Added Ballot serialization tests and fixed bugs

alexsporn committed:
- Remove unused function

alexsporn committed:
- Added serialization tests for participation

alexsporn committed:
- Added serialization tests for Participations

alexsporn committed:
- Added serialization tests for Event

alexsporn committed:
- Set default participation index key to `PARTICIPATE`

muXxer commented on pkg/model/participation/status.go:
```diff
@@ -87,7 +107,7 @@ func (pm *ParticipationManager) EventStatus(eventID EventID) (*EventStatus, erro
 		status.Questions = append(status.Questions, questionStatus)
 	}
-
+	status.Checksum = hex.EncodeToString(statusHash.Sum(nil))
```
Didn't you forget the value of 0 and 255? Is that important?

alexsporn replied:
Look closer, everything happens inside the inner `balanceForAnswerValue` function. That one is used for all values, also for 0 and 255

muXxer replied:
You are right, that looked weird, but only in the github diff. :weirdsmileyguy:

muXxer commented on go.mod:
```diff
@@ -92,7 +92,9 @@ require (
 	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/davidlazar/go-crypto v0.0.0-20200604182044-b73af7476f6c // indirect
 	github.com/dgraph-io/badger v1.6.2 // indirect
+	github.com/dgraph-io/badger/v2 v2.2007.3 // indirect
```
where does the scam come from?

alexsporn replied:
I had to run `go mod tidy` and this came up 🤷‍♂️

muXxer commented on pkg/model/participation/answer.go:
```diff
@@ -2,6 +2,7 @@ package participation
 
 import (
 	"encoding/json"
+	"errors"
```
Normally we use the other errors package.

muXxer replied:
"github.com/pkg/errors"

muXxer commented on pkg/model/participation/participation.go:
```diff
@@ -3,11 +3,16 @@ package participation
 import (
 	"encoding/hex"
 	"encoding/json"
+	"errors"
```
same here

muXxer reviewed (commented):
LGTM

alexsporn committed:
- Replace "errors" with "github.com/pkg/errors" everywhere
